### PR TITLE
Let InMemoryCache return null if cache is empty

### PR DESCRIPTION
**Pull request by jannes-io**
#1 refers to the feature-request repository:
https://github.com/apollographql/apollo-feature-requests/issues/1

### Checklist:

- [x] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests

### CODE REVIEW

**Comment by jannes-io** [comment]
Failing test because of missing field oversight.
It should only catch the error if the query is not present in cache.

I'll investigate further to come to a nicer solution whilst keeping the missing field errors.

**Comment by jannes-io** [comment]
Tests are passing now.

File: src/ApolloClient.ts [hunk 1]
```diff
@@ -374,7 +374,11 @@ export class ApolloClient<TCacheShape> implements DataProxy {
     options: DataProxy.Query<TVariables>,
     optimistic: boolean = false,
   ): T | null {
-    return this.cache.readQuery<T, TVariables>(options, optimistic);
+    try {
+      return this.cache.readQuery<T, TVariables>(options, optimistic);
+    } catch (e) {
+      return null;
```
Comment (benjamn) [inline src/ApolloClient.ts]: A blind `try`-`catch` block that swallows all exceptions cannot the right way to do this. This repository controls the cache as well as the client, so there's no reason we should need a band-aid like this.

File: src/cache/inmemory/inMemoryCache.ts [hunk 2]
```diff
@@ -102,15 +102,15 @@ export class InMemoryCache extends ApolloCache<NormalizedCacheObject> {
...
-            JSON.stringify({ optimistic, rootId, variables }),
+              c.query,
+              // Different watches can have the same query, optimistic
+              // status, rootId, and variables, but if their callbacks are
+              // different, the (identical) result needs to be delivered to
+              // each distinct callback. The easiest way to achieve that
+              // separation is to include c.callback in the cache key for
+              // maybeBroadcastWatch calls. See issue #5733.
+              c.callback,
+              JSON.stringify({ optimistic, rootId, variables }),
```
Comment (benjamn) [inline src/cache/inmemory/inMemoryCache.ts]: Please revert these whitespace changes.

File: src/ApolloClient.ts (hunk 1 above)
Comment (jannes-io) [reply src/ApolloClient.ts]: I fully agree, unfortunately I am not familiar enough with the apollo code-base and just wanted to do a suggestion and raise some attention to the implementation of the requested feature as our code-base for a new project is currently suffering from it

**Comment by EmreErdogan** [comment]
Thanks for the PR and thanks reviewing it. Is this likely to be merged?

**Comment by jannes-io** [comment]
@EmreErdogan no, this would indeed solve the issue, but is not a good way to go about it. Someone with more knowledge about the inMemoryCache will have to take a closer look and come up with a solid fix instead of the band-aid that is this PR.

**Pull request closed without merging**